Mathlib/Algebra/Algebra/Unitize.lean

```diff
@@ -0,0 +1,348 @@
+
+* `NonUnitalSubalgebra.unitization s : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A)`:
+  where `s` is a non-unital subalgebra of a unital `R`-algebra `A`, this is the natural algebra
+  homomorphism sending `(r, a)` to `r • 1 + a`. This is always surjective.#check
+* `NonUnitalSubalgebra.unitizationAlgEquiv s : Unitization R s ≃ₐ[R] Algebra.adjoin R (s : Set A)`:
+  when `R` is a field and `1 ∉ s`, the above homomorphism is injective is upgraded to
+  an `AlgEquiv`.
+* `Subsemiring.closureEquivAdjoinNat : Subsemiring.closure s ≃ₐ[ℕ] Algebra.adjoin ℕ s`: the
+  identity map between these subsemirings, viewed as `ℕ`-algebras.
+* `Subring.closureEquivAdjoinInt : Subring.closure s ≃ₐ[ℤ] Algebra.adjoin ℤ s`: the
+  identity map between these subsemirings, viewed as `ℤ`-algebras.
+* `NonUnitalSubsemiring.unitization : Unitization ℕ S →ₐ[ℕ] Subsemiring.closure (S : Set R)`:
+  the natural `ℕ`-algebra homomorphism between the unitization of a non-unital subsemiring `S` and
+  its `Subsemiring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
+  codomain using `NonUnitalSubsemiring.closureEquivAdjoinNat`
+* `NonUnitalSubring.unitization : Unitization ℤ S →ₐ[ℤ] Subring.closure (S : Set R)`:
+  the natural `ℤ`-algebra homomorphism between the unitization of a non-unital subring `S` and
+  its `Subring.closure`. This is just `NonUnitalSubalgebra.unitization S` where we replace the
+  codomain using `NonUnitalSubring.closureEquivAdjoinInt`
+
+-/
+
+section Generic
+
+variable {R S A : Type _} [CommSemiring R] [Semiring A] [Algebra R A] [SetLike S A]
+  [hSA : NonUnitalSubsemiringClass S A] [hSRA : SMulMemClass S R A] (s : S)
+
+/-- The natural `R`-algebra homomorphism from the unitization of a non-unital subalgebra to
+its `Algebra.adjoin`. -/
+def NonUnitalSubalgebra.unitization : Unitization R s →ₐ[R] Algebra.adjoin R (s : Set A) :=
+  Unitization.lift
+    { toFun := fun x : s => (⟨x, Algebra.subset_adjoin x.prop⟩ : Algebra.adjoin R (s : Set A))
+      map_smul' := fun (_ : R) _ => rfl
+      map_zero' := rfl
+      map_add' := fun _ _ => rfl
+      map_mul' := fun _ _ => rfl }
+
+@[simp]
+theorem NonUnitalSubalgebra.unitization_apply_coe (x : Unitization R s) :
+    (NonUnitalSubalgebra.unitization s x : A) =
+      algebraMap R (Algebra.adjoin R (s : Set A)) x.fst + x.snd :=
+  rfl
+
+theorem NonUnitalSubalgebra.unitization_surjective :
+    Function.Surjective (NonUnitalSubalgebra.unitization s) := by
+  apply Algebra.adjoin_induction'
+  · refine' fun x hx => ⟨(0, ⟨x, hx⟩), Subtype.ext _⟩
+    simp only [NonUnitalSubalgebra.unitization_apply_coe, Subtype.coe_mk]
+    change (algebraMap R { x // x ∈ Algebra.adjoin R (s : Set A) } 0 : A) + x = x
+    rw [map_zero, Subsemiring.coe_zero, zero_add]
+  · exact fun r => ⟨algebraMap R (Unitization R s) r, AlgHom.commutes _ r⟩
+  · rintro _ _ ⟨x, rfl⟩ ⟨y, rfl⟩
+    exact ⟨x + y, map_add _ _ _⟩
+  · rintro _ _ ⟨x, rfl⟩ ⟨y, rfl⟩
+    exact ⟨x * y, map_mul _ _ _⟩
```

Comment:
I see why it's not immediate (you'd have to get back to the full codomain), but is there really no good way of using [Algebra.adjoin_le](https://leanprover-community.github.io/mathlib4_docs/Mathlib/RingTheory/Adjoin/Basic.html#Algebra.adjoin_le) here?

Reply:
Good point, this works:
```lean
  have : Algebra.adjoin R s ≤ ((Algebra.adjoin R (s : Set A)).val.comp (unitization s)).range :=
    Algebra.adjoin_le fun a ha ↦ ⟨(⟨a, ha⟩ : s), by simp⟩
  fun x ↦ match this x.property with | ⟨y, hy⟩ => ⟨y, Subtype.ext hy⟩
```